python/astro_metadata_translator/file_helpers.py

```diff
@@ -0,0 +1,248 @@
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the LICENSE file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# Use of this source code is governed by a 3-clause BSD-style
+# license that can be found in the LICENSE file.
+
+"""Support functions for script implementations."""
+
+__all__ = ("read_fits_metadata", "find_files", "read_basic_metadata_from_file", "read_file_info")
+
+import json
+import re
+import os
+import sys
+import traceback
+
+from astro_metadata_translator import merge_headers, ObservationInfo
+from .tests import read_test_file
+
+
+# Prefer afw over Astropy
+try:
+    from lsst.afw.fits import readMetadata
+    import lsst.daf.base  # noqa: F401 need PropertyBase for readMetadata
+
+    def read_fits_metadata(file, hdu, can_raise=False):
+        """Read a FITS header using afw.
+
+        Parameters
+        ----------
+        file : `str`
+            The file to read.
+        hdu : `int`
+            The header number to read.
+        can_raise : `bool`, optional
+            Indicate whether the function can raise and exception (default)
+            or should return `None` on error. Can still raise if an unexpected
+            error is encountered.
+
+        Returns
+        -------
+        md : `dict`
+            The requested header. `None` if it could not be read and
+            ``can_raise`` is `False`.
+        """
+        try:
+            return readMetadata(file, hdu=hdu)
+        except lsst.afw.fits.FitsError as e:
+            if can_raise:
+                # Try to convert a basic fits error code
+                if "(104)" in str(e):
+                    raise FileNotFoundError(f"No such file or directory: {file}") from e
+                raise e
+            return None
+
+except ImportError:
+    from astropy.io import fits
+
+    def read_fits_metadata(file, hdu, can_raise=False):
+        """Read a FITS header using astropy.
+
+        Parameters
+        ----------
+        file : `str`
+            The file to read.
+        hdu : `int`
+            The header number to read.
+        can_raise : `bool`, optional
+            Indicate whether the function can raise and exception (default)
+            or should return `None` on error. Can still raise if an unexpected
+            error is encountered.
+
+        Returns
+        -------
+        md : `dict`
+            The requested header. `None` if it could not be read and
+            ``can_raise`` is `False`.
+        """
+        header = None
+        try:
+            with fits.open(file) as fits_file:
+                try:
+                    header = fits_file[hdu].header
+                except IndexError as e:
+                    if can_raise:
+                        raise e
+        except Exception as e:
+            if can_raise:
+                raise e
+        return header
```

Comment:
That function has existed like that for ages but shows up as a diff because of the move from one file to another. I added docstrings when I moved and yes it's a bit annoying to repeat oneself. This is really an internal function anyhow. There is no expectation that this is used outside of the command line tooling or test code.

Reply:
Ok, I'm just channeling an old review by John Swinbank where he discouraged me from doing this.